### PR TITLE
[FLINK-28799] Hybrid shuffle should supports schedule the graph contains blocking edge

### PR DESCRIPTION
**Pull request by reswqa**
## What is the purpose of the change

*Based on TPC-DS test, we found that hybrid shuffle can't schedule the graph contains blocking edge. The reason is that some batch operators will forcibly set the exchange mode to blocking, which breaks ALL_ EDGE_HYBRID‘s constraint and may cause scheduling deadlock. We should use a better way to support the scheduling of all kinds of edges, including hybrid edge.*

## Brief change log
  - *Migrate `PipelinedRegionSchedulingStrategyTest` and `StrategyTestUtil` to Junit5 and AssertJ.*
  - *`PipelinedRegionSchedulingStrategy` supports all `ResultPartitionType`.*

## Verifying this change

This change added tests in `PipelinedRegionSchedulingStrategyTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/PipelinedRegionSchedulingStrategy.java [hunk 1]
```diff
@@ -53,10 +53,20 @@ public class PipelinedRegionSchedulingStrategy implements SchedulingStrategy {
     private final Map<SchedulingPipelinedRegion, List<ExecutionVertexID>> regionVerticesSorted =
             new IdentityHashMap<>();
 
+    /** All downstream regions of one schedulingPipelinedRegion. */
+    private final Map<SchedulingPipelinedRegion, Set<SchedulingPipelinedRegion>>
+            downstreamRegionsOfRegion = new IdentityHashMap<>();
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/PipelinedRegionSchedulingStrategy.java:58]: It should be good enough to store a mapping from the upstream region to the downstream consumed partition group, as we already have another map helps mapping the groups to downstream regions.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/PipelinedRegionSchedulingStrategy.java [hunk 2]
```diff
@@ -53,10 +53,20 @@ public class PipelinedRegionSchedulingStrategy implements SchedulingStrategy {
...
+
+    /** All downstream regions of one ExecutionVertexID. */
+    private final Map<ExecutionVertexID, Set<SchedulingPipelinedRegion>> downstreamRegionsOfVertex =
+            new IdentityHashMap<>();
```
[the remark this answers is not in the document]
Comment (reswqa) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/PipelinedRegionSchedulingStrategy.java]: Removed it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/PipelinedRegionSchedulingStrategy.java (hunk 1 above)
Comment (reswqa) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/PipelinedRegionSchedulingStrategy.java]: Changed it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/PipelinedRegionSchedulingStrategy.java [hunk 3]
```diff
@@ -82,14 +90,29 @@ private void init() {
         }
     }
 
+    private void initDownstreamRegionsOfRegionAndVertex() {
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/PipelinedRegionSchedulingStrategy.java:93]:
```suggestion
    private void initDownstreamRegionsOfRegion() {
```

**Review by xintongsong — APPROVED** [review]
Thanks @reswqa. LGTM. I think it would be good to also have @zhuzhurk take a look at it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/PipelinedRegionSchedulingStrategy.java (hunk 3 above)
Comment (reswqa) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/PipelinedRegionSchedulingStrategy.java]: Renamed.

**Pull request closed without merging**